tests.py

```diff
@@ -17,172 +23,183 @@
 torch.manual_seed(42)
 
 
-def test_scaled_dpa():
-    batch_size = 2
-    num_heads = 4
-    seq_length = 5
-    dim_k = 8
+
+BATCH_SIZES = [1, 8, 16, 32]
+NUM_HEADS = [1, 2, 4, 8]
+SEQ_LENGTHS = [10, 50, 100]
+DIM_KS = [8, 16, 32, 64]
+EMB_SIZES = [64, 128, 256, 512]
+DROPOUTS = [0] # can't use other values here since the drop outs occur randomly...
+MAXLEN = [500, 1000, 5000]
+D_FFS = [10, 50, 100, 500]
+NUM_BLOCKS = [1, 2, 4, 8]
+VOCAB_SIZES = [100, 1000, 10000]
+
+@pytest.mark.parametrize("batch_size, num_heads, seq_length, dim_k", list(itertools.product(BATCH_SIZES, NUM_HEADS, SEQ_LENGTHS, DIM_KS)))
+def test_scaled_dpa(batch_size, num_heads, seq_length, dim_k):
+
     query = torch.rand(batch_size, num_heads, seq_length, dim_k)
     key = torch.rand(batch_size, num_heads, seq_length, dim_k)
     value = torch.rand(batch_size, num_heads, seq_length, dim_k)
     mask = torch.ones(batch_size, num_heads, seq_length, seq_length)
 
+    # my implementation
     output, attention_weights = scaled_dpa(query, key, value, mask)
 
-    assert output.shape == (batch_size, num_heads, seq_length, dim_k)
-    assert attention_weights.shape == (batch_size, num_heads, seq_length, seq_length)
+    # builtin pytorch implementation
+    torch_output = F.scaled_dot_product_attention(query, key, value, mask)
+
+    assert torch.allclose(output, torch_output, atol=1e-6), "Output does not match PyTorch's implementation"
+
+
+# the following implementation of positional encodings was taken from https://pytorch.org/tutorials/beginner/translation_transformer.html#seq2seq-network-using-transformer
+class TorchPositionalEncoding(nn.Module):
+    def __init__(self,
+                 emb_size: int,
+                 dropout: float,
+                 maxlen: int = 5000):
+        super(TorchPositionalEncoding, self).__init__()
+        den = torch.exp(- torch.arange(0, emb_size, 2)* math.log(10000) / emb_size)
+        pos = torch.arange(0, maxlen).reshape(maxlen, 1)
+        pos_embedding = torch.zeros((maxlen, emb_size))
+        pos_embedding[:, 0::2] = torch.sin(pos * den)
+        pos_embedding[:, 1::2] = torch.cos(pos * den)
+        pos_embedding = pos_embedding.unsqueeze(-2)
+
+        self.dropout = nn.Dropout(dropout)
+        self.register_buffer('pos_embedding', pos_embedding)
+
+    def forward(self, token_embedding: torch.Tensor):
+        return self.dropout(token_embedding + self.pos_embedding[:token_embedding.size(0), :])
+
+
+@pytest.mark.parametrize("batch_size, seq_len, emb_size, dropout, maxlen", list(itertools.product(BATCH_SIZES, SEQ_LENGTHS, EMB_SIZES, DROPOUTS, MAXLEN)))
+def test_positional_encodings(batch_size, seq_len, emb_size, dropout, maxlen):
+
+    token_emb = torch.rand(batch_size, seq_len, emb_size)
+
+    my_pe = PositionalEncoding(emb_size, dropout, maxlen)
+
+    my_output = my_pe(token_emb)
+
+    torch_pe = PositionalEncoding(emb_size, dropout, maxlen)
```

Comment:
## Comment / Requested Change
* Isn't this supposed to be using the above `TorchPositionalEncoding`?

Reply:
good catch! yes it is.